zirgen/Dialect/ZHLT/Transforms/HoistCommonMuxCode.cpp

```diff
@@ -0,0 +1,178 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "zirgen/Dialect/ZHLT/IR/ZHLT.h"
+#include "zirgen/Dialect/ZHLT/Transforms/PassDetail.h"
+#include "llvm/ADT/Statistic.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "mux-hoisting"
+
+using namespace mlir;
+using namespace zirgen::ZStruct;
+
+namespace zirgen::Zhlt {
+
+namespace {
+
+bool isHoistable(Operation* op) {
+  // Check if all of the operation's operands are defined outside of the mux. If
+  // they are, then we can probably hoist -- but make sure not to reorder side
+  // effects, and never hoist any ops with regions, AliasLayoutOps or block
+  // terminators.
+  return op->getNumRegions() == 0 &&
+         !isa<LoadOp>(op) && // could be more precise by checking for writes within the block
+         !isa<AliasLayoutOp>(op) && !op->hasTrait<OpTrait::IsTerminator>() &&
+         llvm::all_of(op->getOperands(), [=](Value value) {
+           return value.getParentRegion() != op->getParentRegion();
+         });
+}
+
+bool compare(Operation* op1, Operation* op2) {
+  if (op1->getName() != op2->getName() || op1->getNumOperands() != op2->getNumOperands() ||
+      op1->getAttrs().size() != op2->getAttrs().size())
+    return false;
+  for (auto [opn1, opn2] : llvm::zip(op1->getOperands(), op2->getOperands())) {
+    if (opn1 != opn2)
+      return false;
+  }
+  for (auto [attr1, attr2] : llvm::zip(op1->getAttrs(), op2->getAttrs())) {
+    if (attr1 != attr2)
+      return false;
+  }
+  return true;
+}
+
+} // namespace
+
+// Look for code shared between mux arms, and hoist it. If code is common to at
+// least two arms, hoisting it reduces code size. However, this hurts code speed
+// because that code will be unnecessarily executed for any mux arms where it
+// didn't previously occur. That being said, this is always a beneficial
+// optimization if the code occurs on *every* mux arm.
+struct HoistCommonMuxCodePass : public HoistCommonMuxCodeBase<HoistCommonMuxCodePass> {
+  HoistCommonMuxCodePass() = default;
+  HoistCommonMuxCodePass(bool eager) { this->eager = eager; }
+  HoistCommonMuxCodePass(const HoistCommonMuxCodePass& pass) {}
+
+  void runOnOperation() override {
+    getOperation().walk<WalkOrder::PostOrder>([&](SwitchOp mux) {
+      // If code in the mux is shared by multiple but not all mux arms, then
+      // hoisting it reduces code size but increases execution cost. We at worst
+      // break even on execution if it's shared by all mux arms.
+      if (eager) {
+        hoistForSize(mux);
+      } else {
+        hoistForSpeed(mux);
+      }
+    });
+  }
+
+  void hoistForSpeed(SwitchOp mux) {
+    // We're looking for operations shared in all mux arms, and all such
+    // operations must also be in the first mux arm.
+    auto it = mux.getRegion(0).op_begin();
+    while (it != mux->getRegion(0).op_end()) {
+      Operation& op = *(it++);
```

Comment:
This looks to my eyes like a decomposed for loop, and I wonder whether iterating over `Region::getOps` wouldn't be simpler? Perhaps there's a good reason it's done this way, but a comment explaining the unusual structure might be nice.

Reply:
Because hoisting moves operations out of arm 0, this iteration scheme avoids losing track of which operation in the region we're looking at, so using a for loop doesn't work.